app/assets/javascripts/components/box_batches_form.js.jsx

```diff
@@ -0,0 +1,360 @@
+var BoxBatchesForm = React.createClass({
+  getInitialState: function () {
+    var genRandomKey = this.__genRandomKey;
+
+    var batches = this.props.batches || [];
+    batches.forEach(function (batch) {
+      batch.key = genRandomKey(batches);
+    });
+
+    return {
+      batches: batches,
+      showBatchSelector: batches.length == 0,
+    };
+  },
+
+  reset: function () {
+    this.setState({
+      batches: [],
+      showBatchSelector: true,
+    });
+  },
+
+  render: function () {
+    var batches = this.state.batches;
+    var removeBatch = this.removeBatch
+    var onChange = this.onChange
+
+    return (
+      <div className="batches-selector" data-isvalid={this.isValid()}>
+        <a className="clear-batches" href="#" onClick={this.reset} style={{ display: "none" }}></a>
+        <div className="items-count">
+          <div className="title">{batches.length}&nbsp;{batches.length == 1 ? "batch" : "batches"}</div>
+        </div>
+        {batches.map(function (batch) {
+          return <BoxBatchForm
+            key={"box_batches_" + batch.key}
+            batch={batch}
+            onChange={onChange}
+            onRemove={removeBatch}
+          />
+        })}
+        {this.state.showBatchSelector ? this.renderBatchSelector() : this.renderAddBatchButton() }
+      </div>
+    );
+  },
+
+  renderBatchSelector: function () {
+    return (
+      <div className="list-items">
+        <div className="items-row">
+          <CdxSelectAutocomplete
+            className="input-block"
+            url={this.props.findBatchUrl}
+            name="box[batch_uuids]"
+            placeholder="Enter batch id"
+            value=""
+            prepareOptions={this.prepareOptions}
+            autoselect={true}
+            onSelect={this.selectBatch}
+          />
+        </div>
+      </div>
+    );
+  },
+
+  renderAddBatchButton: function () {
+    return (
+      <a className="add-items" href="#" onClick={this.showBatchSelector}>
+        <div className="add-items">
+          <div className="icon-circle-plus icon-blue icon-margin"></div>
+          <div className="add-link">ADD BATCH</div>
+        </div>
+      </a>
+    );
+  },
+
+  showBatchSelector: function (event) {
+    event.preventDefault();
+    this.setState({ showBatchSelector: true });
+  },
+
+  selectBatch: function (_, options) {
+    if (options.length == 0) return;
+
+    var batches = this.state.batches;
+    var genRandomKey = this.__genRandomKey;
+
+    var newBatches = options.map(function (option) {
+      return {
+        key: genRandomKey(batches),
```

Comment:
I can't use the batch UUID, because the current form allows to set the same batch twice (or more). One valid reason is that there could be different instructions. I thus use a randomly generated key to make sure that React will correctly diff/patch/update the DOM.